lib/parser/provenance.cc

```diff
@@ -149,15 +149,20 @@ ProvenanceRange AllSources::AddCompilerInsertion(std::string text) {
   return covers;
 }
 
-void AllSources::EmitMessage(std::ostream &o, ProvenanceRange range,
-    const std::string &message, bool echoSourceLine) const {
-  CHECK(IsValid(range));
-  const Origin &origin{MapToOrigin(range.start())};
+void AllSources::EmitMessage(std::ostream &o,
+    const std::optional<ProvenanceRange> &range, const std::string &message,
```

Comment:
Now we can emit messages that aren't associated with a source location. Should there be a way to create such messages? For example, in `ModFileWriter` errors are saved as `vector<MessageFormattedText>` because they have no source location. It would be nice to be able to treat them as normal `Messages`.

Reply:
There's a mechanism in place already for describing text in the AllSources that didn't come from anywhere with a meaningful location, such as the occasional spaces or punctuation that get inserted by preprocessing.  Provenances that map to such text are left empty when messages that have such provenances are formatted.  We might be able to use that facility in this situation.